core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -233,17 +235,13 @@ private[spark] class MemoryStore(
     }
 
     if (keepUnrolling) {
-      // We successfully unrolled the entirety of this block
-      val arrayValues = vector.toArray
-      vector = null
-      val entry =
-        new DeserializedMemoryEntry[T](arrayValues, SizeEstimator.estimate(arrayValues), classTag)
-      val size = entry.size
+      // get the precise size
+      val size = estimateSize(true)
```

Comment:
Why we need `estimateSize(true)`? Is this just creating the entry and getting `entry.size`

Reply:
This is just we unrolled the iterator successfully. But maybe the size of underlying vector is greater than `unrollMemoryUsedByThisBlock `, so we need request more memory. In this time, there is a possible that we can't request enough memory again, so we should call `bbos.toChunkedByteBuffer` or `vector.toArray` after requested enough memory.

Reply:
We just unrolled the iterator successfully until here. But the size of underlying vector maybe greater than the `unrollMemoryUsedByThisBlock` which we requested memory for unroll the block. So we need check it again and determine whether we need request more memory. And we only should call `bbos.toChunkedByteBuffer`  or `vector.toArray` after we requested enough memory.

Reply:
But the previous code just calls `entry.size`, are you fixing a new bug?

Reply:
Previously, the `putIteratorAsValues ` seems no problem. But the `putIteratorAsBytes ` doesn't check again after unrolled the iterator.  Now the `putIterator` is copied form previous `putIteratorAsValues `. For `SizeTrackingVector`, we could call `arrayValues.toIterator` to get a iterator again after call `SizeTrackingVector.toArray`. But for `ChunkedByteBufferOutputStream`, we can't back to `stream` after called `ChunkedByteBufferOutputStream.toChunkedByteBuffer` (the `PartiallySerializedBlock` need a stream).